### PR TITLE
*: rewrite builtin function ROW && remove `Datum.KindRow`

### PR DESCRIPTION
**Pull request by XuHuaiyu**
to #4080 
PTAL @zz-jason @shenli @jackysp

### CODE REVIEW

**Comment by zz-jason** [comment]
to #4336

**Comment by zz-jason** [comment]
@XuHuaiyu ci failed.

**Comment by hanfei1991** [comment]
When do we still need row function ?

File: expression/builtin_other.go [hunk 1]
```diff
@@ -46,29 +46,27 @@ type rowFunctionClass struct {
...
-func (b *builtinRowSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return types.Datum{}, errors.Trace(err)
-	}
-	d.SetRow(args)
-	return
+// rowFunc should always be flattened in expression rewrite phrase.
+func (b *builtinRowSig) evalString(row []types.Datum) (string, bool, error) {
+	return "", false, errors.New("`builtinRowSig.evalString()` should never be called.")
```
Comment (jackysp) [inline expression/builtin_other.go]: Why not return error directly in `getFunction`?

Comment (XuHuaiyu) [reply expression/builtin_other.go]: We still need rowFunc now, but we do not support the evaluation of it.

**Comment by hanfei1991** [comment]
LGTM

**Comment by XuHuaiyu** [comment]
@zz-jason  PTAL

File: expression/builtin_other.go [hunk 2]
```diff
@@ -46,29 +46,27 @@ type rowFunctionClass struct {
...
-func (b *builtinRowSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return types.Datum{}, errors.Trace(err)
-	}
-	d.SetRow(args)
-	return
+// rowFunc should always be flattened in expression rewrite phrase.
+func (b *builtinRowSig) evalString(row []types.Datum) (string, bool, error) {
+	return "", false, errors.New("should never be called")
```
Comment (zz-jason) [inline expression/builtin_other.go]: how about change the error message to a more specific one, like "evaluation of function ROW should never be called ?".

maybe we can just panic here ?

**Comment by XuHuaiyu** [comment]
PTAL @zz-jason

**Comment by zz-jason** [comment]
@XuHuaiyu ci failed.

**Pull request merged**